src/remoteConnector.ts

```diff
@@ -900,12 +903,76 @@ export default class RemoteConnector extends Disposable {
 		}
 	}
 
-	private startHeartBeat(accessToken: string, connectionInfo: SSHConnectionParams) {
+	private async startHeartBeat(accessToken: string, connectionInfo: SSHConnectionParams, gitpodVersion: GitpodVersion) {
 		if (this.heartbeatManager) {
 			return;
 		}
 
 		this.heartbeatManager = new HeartbeatManager(connectionInfo.gitpodHost, connectionInfo.workspaceId, connectionInfo.instanceId, accessToken, this.logger, this.telemetry);
+
+		// gitpod remote extension installation is async so sometimes gitpod-desktop will activate before gitpod-remote
+		// let's try a few times for it to finish install
+		try {
+			await retry(async () => {
+				await vscode.commands.executeCommand('__gitpod.cancelGitpodRemoteHeartbeat');
+			}, 3000, 15);
+			this.telemetry.sendTelemetryEvent('vscode_desktop_heartbeat_state', { enabled: String(true), gitpodHost: connectionInfo.gitpodHost, workspaceId: connectionInfo.workspaceId, instanceId: connectionInfo.instanceId, gitpodVersion: gitpodVersion.raw });
+		} catch {
+			this.logger.error(`Could not execute '__gitpod.cancelGitpodRemoteHeartbeat' command`);
+			this.telemetry.sendTelemetryEvent('vscode_desktop_heartbeat_state', { enabled: String(false), gitpodHost: connectionInfo.gitpodHost, workspaceId: connectionInfo.workspaceId, instanceId: connectionInfo.instanceId, gitpodVersion: gitpodVersion.raw });
+		}
+	}
+
+	private async initializeRemoteExtensions() {
+		let syncData: { ref: string; content: string } | undefined;
+		try {
+			syncData = await this.settingsSync.readResource(SyncResource.Extensions);
```

Comment:
I wonder should not we rather provider gitpodHost here from connectionInfo instead of reading from settings. I think for remote window connectionInfo is source of the truth.

Reply:
We were discussing yesterday with self hosting team that having different configurations per host will be good, i.e. one should be able to specify different values for different hosts. Like use SSH Gateway for gitpod.io, but local app for SH solution and so on. How we designed it with one host does not really fit to it.